csrc/ir/nodes.cpp

```diff
@@ -4540,8 +4541,33 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
   const auto& [sizes, strides] = inferShapeOfOutput(out(), ee);
   auto meta_out = at::detail::empty_strided_meta(sizes, strides, a.dtype());
   if (meta_out.is_contiguous()) {
-    return {matmul_out};
+    // When the contracting dimension is sharded, each device has a partial
+    // matmul output and is followed by an allreduce. For loop split, this is
+    // represented as an rfactored reduction. The local matmul logical domain
+    // after the rfactor is: i{DIDx}, i{M}, i{N}, r{K//d}. Unsqueeze the
+    // rfactored DID axis to correctly bind with the logical domain. See
+    // tests/python/test_multidevice.py/test_matmul_allreduce_loop_split
+    std::vector<int64_t> rfactor_did_idx;
+    auto out_logical = out()->getLogicalDomain();
+    for (auto idx : c10::irange(out_logical.size())) {
+      if (out_logical.at(idx)->isRFactorProduct() &&
+          out_logical.at(idx)->isDeviceDim()) {
+        rfactor_did_idx.push_back(idx);
+      }
+    }
+    if (rfactor_did_idx.empty()) {
+      return {matmul_out};
+    }
+
+    NVF_ERROR(
+        rfactor_did_idx.size() == 1,
+        "Expected only 1 rfactored DID iterdomain, got: ",
+        rfactor_did_idx.size());
+    return {matmul_out.unsqueeze(rfactor_did_idx.front())};
   }
+  NVF_CHECK(
+      !(isSharded(out())),
```

Comment:
This check appears too strong. I don't think there's a problem to copy matmul_out, a local shard of a distributed tensor, to strided_matmul_out.

Reply:
If you agree with that, then the logic of the whole function can be simplified to: 

```
collect rfactor_did_idx
if (rfactor_did_idx != -1) {
  matmul_out.unsqueeze(rfactor_did_idx);
}

if (meta_out is contiguous) {
  return matmul_out;
}

strided_matmul_out = ...
copy matmul_out 
return strided_matmul_out;
```

Reply:
In that case, I should also move the unsqueeze for rf DIDx outside the if block.